File: java/Tabel.java
import java.util.List;

public class Tabel {

    public static void printDashedLine(int... widths) {
        System.out.print("+");
        for (int width : widths) {
            for (int i = 0; i < width + 2; i++) { // Add 2 for side borders
                System.out.print("-");
            }
            System.out.print("+");
        }
        System.out.println();
    }

    public static void printProducts(List<Product> products) {
        if (products.isEmpty()) {
            System.out.println("No products to display.");
            return;
        }

        // Initialize initial column widths for each attribute
        int idWidth = 3;
        int nameWidth = 4;
        int brandWidth = 5;
        int priceWidth = 5;
        int sizeWidth = 4;
        int materialWidth = 8;
        int genderWidth = 6;
        int colorWidth = 5;
        int sleeveTypeWidth = 11;

// Calculate maximum width for each attribute
for (Product product : products) {
    if (product instanceof Shirt) {
        Shirt shirt = (Shirt) product;
        idWidth = Math.max(idWidth, String.valueOf(shirt.idProduct).length() + 2);
        nameWidth = Math.max(nameWidth, shirt.name.length() + 2);
        brandWidth = Math.max(brandWidth, shirt.brand.length() + 2);
        int actualPriceWidth = ("Rp. " + shirt.price).length() + 2; // 2 for "Rp." prefix
        priceWidth = Math.max(priceWidth, actualPriceWidth);
        sizeWidth = Math.max(sizeWidth, shirt.size.length() + 2);
        materialWidth = Math.max(materialWidth, shirt.material.length() + 2);
        genderWidth = Math.max(genderWidth, shirt.gender.length() + 2);
        colorWidth = Math.max(colorWidth, shirt.color.length() + 2);
        sleeveTypeWidth = Math.max(sleeveTypeWidth, shirt.sleeveType.length() + 2);
    }
}

        // Print header row
        printDashedLine(idWidth, nameWidth, brandWidth, priceWidth, sizeWidth, materialWidth, genderWidth, colorWidth, sleeveTypeWidth);
        System.out.printf("| %-" + idWidth + "s | %-" + nameWidth + "s | %-" + brandWidth + "s | %-" + priceWidth + "s | %-" + sizeWidth + "s | %-" + materialWidth + "s | %-" + genderWidth + "s | %-" + colorWidth + "s | %-" + sleeveTypeWidth + "s |%n",
                "ID", "Name", "Brand", "Price", "Size", "Material", "Gender", "Color", "Sleeve Type");
        printDashedLine(idWidth, nameWidth, brandWidth, priceWidth, sizeWidth, materialWidth, genderWidth, colorWidth, sleeveTypeWidth);

        // Print product data
        for (Product product : products) {
            if (product instanceof Shirt) {
                Shirt shirt = (Shirt) product;
                System.out.printf("| %-" + idWidth + "d | %-" + nameWidth + "s | %-" + brandWidth + "s | Rp.%-" + (priceWidth - 3) + "s | %-" + sizeWidth + "s | %-" + materialWidth + "s | %-" + genderWidth + "s | %-" + colorWidth + "s | %-" + sleeveTypeWidth + "s |%n",
                shirt.idProduct, shirt.name, shirt.brand, shirt.price, shirt.size, shirt.material, shirt.gender, shirt.color, shirt.sleeveType);
    }
        }

        // Print dashed line at the end of the table
        printDashedLine(idWidth, nameWidth, brandWidth, priceWidth, sizeWidth, materialWidth, genderWidth, colorWidth, sleeveTypeWidth);
    }
}

File: py/Tabel.py
from typing import List
from Product import Product
from Shirt import Shirt

class TablePrinter:
    
    def print_products(products: List[Product]):
        if not products:
            print("Tidak ada produk untuk ditampilkan.")  # Menampilkan pesan jika tidak ada produk
            return

        # Menemukan panjang maksimum untuk setiap kolom
        id_length = max(len(str(product.idProduct)) for product in products)
        name_length = max(len(product.name) for product in products)
        brand_length = max(len(product.brand) for product in products)
        price_length = max(len(product.price) for product in products)
        size_length = max(len(product.size) for product in products)
        material_length = max(len(product.material) for product in products)
        gender_length = max(len(product.gender) for product in products)
        color_length = max(len(product.color) for product in products)
        sleeve_type_length = max(len(product.sleeveType) for product in products)

        # Menyesuaikan panjang kolom agar sesuai dengan panjang header
        id_length = max(id_length, len("ID"))
        name_length = max(name_length, len("Nama"))
        brand_length = max(brand_length, len("Merek"))
        price_length = max(price_length, len("Harga"))
        size_length = max(size_length, len("Ukuran"))
        material_length = max(material_length, len("Bahan"))
        gender_length = max(gender_length, len("Jenis Kelamin"))
        color_length = max(color_length, len("Warna"))
        sleeve_type_length = max(sleeve_type_length, len("Jenis Lengan"))

        # Mencetak header tabel
        print(f"+{'-'*(id_length+2)}+{'-'*(name_length+2)}+{'-'*(brand_length+2)}+{'-'*(price_length+2)}+{'-'*(size_length+2)}+{'-'*(material_length+2)}+{'-'*(gender_length+2)}+{'-'*(color_length+2)}+{'-'*(sleeve_type_length+2)}+")
        print(f"| {'ID':<{id_length}} | {'Nama':<{name_length}} | {'Merek':<{brand_length}} | {'Harga':<{price_length}} | {'Ukuran':<{size_length}} | {'Bahan':<{material_length}} | {'Jenis Kelamin':<{gender_length}} | {'Warna':<{color_length}} | {'Jenis Lengan':<{sleeve_type_length}} |")
        print(f"+{'-'*(id_length+2)}+{'-'*(name_length+2)}+{'-'*(brand_length+2)}+{'-'*(price_length+2)}+{'-'*(size_length+2)}+{'-'*(material_length+2)}+{'-'*(gender_length+2)}+{'-'*(color_length+2)}+{'-'*(sleeve_type_length+2)}+")

        # Mencetak data produk
        for product in products:
            if isinstance(product, Shirt):
                print(f"| {product.idProduct:<{id_length}} | {product.name:<{name_length}} | {product.brand:<{brand_length}} | {product.price:<{price_length}} | {product.size:<{size_length}} | {product.material:<{material_length}} | {product.gender:<{gender_length}} | {product.color:<{color_length}} | {product.sleeveType:<{sleeve_type_length}} |")

        # Mencetak garis putus-putus di akhir tabel
        print(f"+{'-'*(id_length+2)}+{'-'*(name_length+2)}+{'-'*(brand_length+2)}+{'-'*(price_length+2)}+{'-'*(size_length+2)}+{'-'*(material_length+2)}+{'-'*(gender_length+2)}+{'-'*(color_length+2)}+{'-'*(sleeve_type_length+2)}+")
